pinot-core/src/main/java/org/apache/pinot/core/data/table/IndexedTableResizer.java

```diff
@@ -0,0 +1,307 @@
+      String aggregationColumn = AggregationFunctionUtils.getAggregationColumnName(aggregationInfo);
+      aggregationColumnToIndex.put(aggregationColumn, i);
+      aggregationColumnToInfo.put(aggregationColumn, aggregationInfo);
+    }
+
+    _numOrderBy = orderBy.size();
+    _orderByValueExtractors = new OrderByValueExtractor[_numOrderBy];
+    Comparator[] comparators = new Comparator[_numOrderBy];
+
+    for (int i = 0; i < _numOrderBy; i++) {
+      SelectionSort selectionSort = orderBy.get(i);
+      String column = selectionSort.getColumn();
+
+      if (keyIndexMap.containsKey(column)) {
+        int index = keyIndexMap.get(column);
+        _orderByValueExtractors[i] = new KeyColumnExtractor(index);
+      } else if (aggregationColumnToIndex.containsKey(column)) {
+        int index = aggregationColumnToIndex.get(column);
+        AggregationInfo aggregationInfo = aggregationColumnToInfo.get(column);
+        AggregationFunction aggregationFunction =
+            AggregationFunctionUtils.getAggregationFunctionContext(aggregationInfo).getAggregationFunction();
+        _orderByValueExtractors[i] = new AggregationColumnExtractor(index, aggregationFunction);
+      } else {
+        throw new IllegalStateException("Could not find column " + column + " in data schema");
+      }
+
+      comparators[i] = Comparator.naturalOrder();
+      if (!selectionSort.isIsAsc()) {
+        comparators[i] = comparators[i].reversed();
+      }
+    }
+
+    _intermediateRecordComparator = (o1, o2) -> {
+
+      for (int i = 0; i < _numOrderBy; i++) {
+        int result = comparators[i].compare(o1._values[i], o2._values[i]);
+        if (result != 0) {
+          return result;
+        }
+      }
+      return 0;
+    };
+  }
+
+  /**
+   * Constructs an IntermediateRecord from Record
+   * The IntermediateRecord::key is the same Record::key
+   * The IntermediateRecord::values contains only the order by columns, in the query's sort sequence
+   * For aggregation values in the order by, the final result is extracted if the intermediate result is non-comparable
+   */
+  @VisibleForTesting
+  IntermediateRecord getIntermediateRecord(Record record) {
+    Comparable[] intermediateRecordValues = new Comparable[_numOrderBy];
+    for (int i = 0; i < _numOrderBy; i++) {
+      intermediateRecordValues[i] = _orderByValueExtractors[i].extract(record);
+    }
+    return new IntermediateRecord(record.getKey(), intermediateRecordValues);
+  }
+
+  /**
+   * Trim recordsMap to trimToSize, based on order by information
+   * Resize only if number of records is greater than trimToSize
+   * The resizer smartly chooses to create PQ of records to evict or records to retain, based on the number of records and the number of records to evict
+   */
+  void resizeRecordsMap(Map<Key, Record> recordsMap, int trimToSize) {
+
+    int numRecordsToEvict = recordsMap.size() - trimToSize;
+
+    if (numRecordsToEvict > 0) {
+      int size;
```

Comment:
(nit) no need to declare these variables outside of the if clause. Easier to read if declared inside the if clause IMO